### DISCUSSION TITLE
Ewrapper interface equivalent?

### DISCUSSION THREAD

**Discussion by KeithKee** [wboayue/rust-ibapi#186] [General]
Many thanks for sharing your code. I am a rust and python newbie, having implemented some working TWSAPI python for receiving tick data for SPX and SPY.  Many of the examples I tried just end with no meaningful messages. The eclient connection should receive all messages from TWS. I would be very grateful if you could share how the EWrapper interface is implemented in rust-ibapi?

**Comment by wboayue** [comment]
`rust-ibapi` does not implement the `EWrapper` interface. Instead, interaction with the API is primarily done through the `Client`. Errors that occur during a request are returned as the result of the corresponding call.

For system messages not tied to a specific request, these are currently sent to the log file. To capture these messages, ensure that a logging system is initialized. If you are using `env_logger`, you can initialize it with the following line:

```rust
 env_logger::init();
```

Then, set the `RUST_LOG` environment variable to your desired log level (e.g., `debug`).

That said, it may be worth considering an enhancement to allow users to register a callback on the Client for handling system messages directly.

**Reply by KeithKee** [reply]
Many thanks for your quick clear reply.

**Reply by SpekkleDust** [reply]
@wboayue excellent piece of work. I do feel the eWrapper is somewhat fundamental. Any plans to include this?

**Reply by wboayue** [reply]
I’m aware that `EWrapper` is a fundamental part of many IB API clients, and I understand why a callback-driven paradigm could be appealing. However, in this library, I’ve chosen a design that keeps responses explicitly tied to their originating requests. This approach can make the application logic more transparent especially for users who find direct mappings from requests to responses more intuitive than a global callback interface. Since the official and other third-party libraries already implement `EWrapper` style callbacks, I believe my library can add more value by focusing on a different design philosophy. That said, I appreciate the feedback and the chance to clarify the library’s goals.
